File: scala/src/Problem_003.scala

/**
 * What is the largest prime factor of the number 600851475143?
 */
object Problem_003 extends App {
  import Predef.timed

  def maxPrimeFactor(n: Long) = {
    def divide(n: Long,  fac: Int): Int = {
      if (n == fac) fac
      else if (n % fac == 0) divide(n / fac, fac)
      else divide(n, fac + 1)
    }

    divide(n, 2)
  }
  
  val duration = timed(assert(6857 == maxPrimeFactor(600851475143L)))
  println("duration: " + duration + " ms.")
}

File: haskell/src/Problem_003.hs

main = print $ maxPrimeFactor 600851475143

maxPrimeFactor n = divide n 2 where
  divide n fac
    | n == fac = fac
    | mod n fac == 0 = divide (div n fac) fac
    | otherwise = divide n (succ fac)
